osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/domain/commands/monitoring/AbstractGetPowerQualityProfileHandler.java

```diff
@@ -463,39 +439,105 @@ private ProfileEntryValueDto makeNumericProfileEntryValueDto(final DataObject da
 
-        final List<ScalerUnitInfo> result = new ArrayList<>();
+        Map<Integer, CaptureObjectDefinitionDto> selectableCaptureObjects = new HashMap<>();
 
+        // there is always only one GetResult
         for (final GetResult captureObjectResult : captureObjects) {
-            final DataObject dataObject = captureObjectResult.getResultData();
-            final List<DataObject> dataObjectList1 = dataObject.getValue();
-            for (final DataObject captureObjectDataObject : dataObjectList1) {
+
+            final List<DataObject> dataObjects = captureObjectResult.getResultData().getValue();
+            int positionInDataObjectsList = 0;
+
+            for (DataObject dataObject : dataObjects) {
 
                 final CosemObjectDefinitionDto cosemObjectDefinitionDto = this.dlmsHelper
-                        .readObjectDefinition(captureObjectDataObject, CAPTURE_OBJECT);
-                final int classId = cosemObjectDefinitionDto.getClassId();
+                        .readObjectDefinition(dataObject, CAPTURE_OBJECT);
+
                 final String logicalName = cosemObjectDefinitionDto.getLogicalName().toString();
-                if (this.hasScalerUnit(classId)) {
-                    final AttributeAddress addr = new AttributeAddress(classId, logicalName,
-                            SCALER_UNITS_MAP.get(classId));
-                    final List<GetResult> scalerUnitResult = this.dlmsHelper
-                            .getAndCheck(conn, device, "retrieve scaler unit for capture object", addr);
-                    final DataObject scalerUnitDataObject = scalerUnitResult.get(0).getResultData();
-                    result.add(new ScalerUnitInfo(logicalName, classId, scalerUnitDataObject));
-                } else {
-                    result.add(new ScalerUnitInfo(logicalName, classId, null));
+
+                if (logicalNames.contains(logicalName)) {
+                    selectableCaptureObjects.put(positionInDataObjectsList,
+                            new CaptureObjectDefinitionDto(cosemObjectDefinitionDto.getClassId(),
+                                    new ObisCodeValuesDto(logicalName),
+                                    (byte) cosemObjectDefinitionDto.getAttributeIndex(),
+                                    cosemObjectDefinitionDto.getDataIndex()));
                 }
+                positionInDataObjectsList++;
+            }
+        }
+
+        return selectableCaptureObjects;
+    }
+
+    private List<ScalerUnitInfo> createScalerUnitInfos(final DlmsConnectionManager conn, final DlmsDevice device,
+            final List<GetResult> captureObjects) throws ProtocolAdapterException {
+
+        List<ScalerUnitInfo> scalerUnitInfos = new ArrayList<>();
+
+        // there is always only one GetResult
+        for (final GetResult captureObjectResult : captureObjects) {
+
+            final List<DataObject> dataObjects = captureObjectResult.getResultData().getValue();
```

Comment:
Potential NullPointerException. The comment suggests this will never happen, but I don't know if it never can happen.

Reply:
I tried with data en without getting data from the meter. so I think we are good. I do understand your point